File: 3838-path-existence-queries-in-a-graph-i/path-existence-queries-in-a-graph-i.cpp
class Solution {
public:
    int find(int x, vector<int> &parent){
        if (parent[x] != x) parent[x] = find(parent[x], parent);
        return parent[x];

    }
    vector<bool> pathExistenceQueries(int n, vector<int>& nums, int maxDiff, vector<vector<int>>& queries) {
        vector<int> parent(n, 0);

        for(int i = 0; i<n; i++) parent[i] = i;

        int pu, pv;

        for(int i = 1; i<n; i++){
            pu = find(i-1, parent);
            pv = find(i, parent);

            if (nums[i] - nums[i-1] <= maxDiff) parent[pu] = parent[pv];
        }

        int u, v;

        vector<bool> ans;

        for(auto &it: queries){
            u = it[0];
            v = it[1];

            if (find(u, parent) == find(v, parent)) ans.push_back(true);
            else ans.push_back(false);
        }

        return ans;
    }
};

File: 3838-path-existence-queries-in-a-graph-i/path-existence-queries-in-a-graph-i.py
class Solution:
    def pathExistenceQueries(self, n: int, nums: List[int], maxDiff: int, queries: List[List[int]]) -> List[bool]:

        # def find(x):
        #     if parent[x]!=x:
        #         parent[x] = find(parent[x])

        #     return parent[x]

        # parent = [i for i in range(n)]
        parent = [0]*n

        # for i in range(1, n):
        #     if nums[i]<=nums[i-1] + maxDiff:
                
        #         pu = find(i-1)
        #         pv = find(i)

        #         parent[pv] = pu

        cid = 0

        for i in range(1, n):
            if nums[i] > nums[i-1] + maxDiff:
                cid+=1

            parent[i] = cid

        answer = []

        for i in queries:
            u = i[0]
            v = i[1]
            
            answer.append(parent[u] == parent[v])

        return answer

        # print(parent)
